### PR TITLE
[reland] Move unboxed dispatch decision into dispatcher

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29337 Hide the OperatorKernel* argument from the stack based kernel API
* #29330 OperatorHandle::callBoxed/callUnboxed
* #29214 Test cases for backend fallback kernels
* #29201 Add a OperatorHandle argument to boxed kernels
* #28494 Implement backend level fallback for c10
* #28314 Simplify c10 dispatcher
* **#29200 [reland] Move unboxed dispatch decision into dispatcher**

Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D18282747](https://our.internmc.facebook.com/intern/diff/D18282747/)

### CODE REVIEW

**Comment by ezyang** [comment]
I'll give you an approve in a little bit, but one thing I am interested to know is how much this diff inflates binary size. After all, it was the diff that pushed us over the edge previously on Windows linking; could it be that actually a lot more code is being generated after this patch? Can we measure this?

**Comment by smessmer** [comment]
@ezyang good catch. Thanks for insisting. This grows the libtorch.so size by 2MB. I'm working on fixing it.

**Comment by smessmer** [comment]
binary size is fixed

File: aten/src/ATen/core/dispatch/DispatchTable.h [hunk 1]
```diff
@@ -189,11 +189,10 @@ class DispatchTable final {
...
    */
-   const KernelFunction& lookupBoxed(const Stack* stack) const {
-     return lookup_([&] () -> c10::optional<TensorTypeId> {
-       return dispatch_strategy_.get_dispatch_key_boxed(stack);
-     });
-   }
+  const KernelFunction& lookupBoxed(const Stack* stack) const {
+    c10::optional<TensorTypeId> dispatchKey = dispatch_strategy_.get_dispatch_key_boxed(stack);
+    return lookup_(dispatchKey);
+  }
```
Comment (ezyang) [inline aten/src/ATen/core/dispatch/DispatchTable.h]: Interesting. Do you know what happened here?

**Review by ezyang — APPROVED** [review]
It would be nice to get some binary size instrumentation *specifically* for the registration gadgets, as the system is complex enough that you cannot easily eyeball how big the binary size of it is.

File: aten/src/ATen/core/dispatch/DispatchTable.h (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/core/dispatch/DispatchTable.h]: This one wasn't the important one. The `lookupUnboxed` and `get_dispatch_key_unboxed` was. They're templates and if it's not short enough to be inlined, they get instantiated for each operator. Same for the lambda inside `lookupUnboxed` which is also a function that is instantiated for each op if it isn't inlined.

**Pull request closed without merging**